### ISSUE TITLE
Bookmap <part> with no href or navtitle generates empty TOC entry in PDF [DOT 2.1 M1]

### ISSUE THREAD

**Issue by raducoravu** [dita-ot/dita-ot#1930]
I have in my bookmap a construct like:

    <part collection-type="sequence">
        <chapter href="topics/installing.dita">
            .............
        </chapter>
    </part>

When producing PDF using DITA OT 1.8 in PDF both the left outline and the table of contents inside the content treat this part as transparent, it does not provide an extra level in the TOC.

But when producing PDF using DITA OT 2.1, this empty part appears as a level in the TOC (an empty level, with just an expansion mark).

We use this kind of construct just to set common attributes on the topicrefs which are inside the part.

**Comment by drmacro** [comment]
I don't think that's an intended use of <part>. A part represents a titled division of a publication, so not having a title (either a navigation title or referenced topic) is semantically wrong.

Either the processor should report this case as an error and/or it should recover as Radu wants (that is, treat the <part> as a topicgroup).

**Comment by raducoravu** [comment]
I also agree that maybe we should not have done this.
Problem with bookmaps is that you cannot add topicgroups and other usual map elements like mapref in any place.

**Comment by drmacro** [comment]
There are many problems with bookmaps, of which that is only one :-)

**Comment by raducoravu** [comment]
In the meantime I modified our user's manual to avoid these constructs. 
Jarno, if you consider that this is not really a bug, just a change in behavior you can also close this one.

**Comment by jelovirt** [comment]
I think this is a bug in OT and an empty part title should be generated.

**Comment by robander** [comment]
@jelovirt @raducoravu Just noticed this one while searching for other issues related to bookmap container elements and wanted your thoughts.

Currently, if I have the construct empty `<part>`, I get "Part I" in the TOC and in the content, but an empty bookmark. 

Normally default processing doesn't include the "Part X" or "Chapter #" in the bookmarks, but I think that would make good default text for this case, since there is no other text to work with. Does that make sense?

**Comment by raducoravu** [comment]
I would treat all chapters, parts without href or keyref as we treat tgroups, ignore the extra level from the TOC, bookmarks and content.

**Comment by robander** [comment]
Taking another look at this, after noticing it again in Radu's list of important backlog items.

My thought for parts:
- If there is a part container with no href and no title, ignore it and move on to chapters -- for TOC, bookmarks, and content. This is a change from current behavior (probably appropriate as a fix for 3.1 but not for 3.0).
- If *any* part in the bookmap _does_ have a title or topic, then I assume this part is there for some reason, and we should keep the current "Part X" in the TOC / content -- and probably match that "Part X" in the TOC.
- I don't really want to touch chapters -- I don't think we should ignore chapters even if they don't have title / topic. That is, if you're using a bookmap then one of the central goals is to have chapters, so it doesn't make sense to have non-chapter top-level topics as part of the hierarchy. Where @drmacro suggested above that parts without title / topic are not an intended use, I think the grouping Radu provided is at least an understandable use -- but I don't really see even that case as valid for chapters.

I know I said in a meeting today that I should focus on higher priority 3.0 issues but I'd already been thinking about this one so wanted to get my thoughts down before I returned to that. Attaching the sample files I'd put together for testing: 
[1930.zip](https://github.com/dita-ot/dita-ot/files/1604261/1930.zip)

**Comment by drmacro** [comment]
I think that if you want a part whose title is "Part 1" then you should specify an empty navigation title.

That is, there's a clear and significant difference between a topicref that has no navigation title (and no referenced topic) and a topicref that has a navigation title, even if that title is the empty string.

So I think it would be incorrect for a processor to generate "Part 1" for a part topicref with no referenced topic and no navigation title.

The BookMap specification could have stated that the semantic of part in this case is to generate a title (that is, that part always acts like a reference to a topic or a topichead even if there is no explicit navtitle, but it did not do so. Therefore, I don't see how processors have license to infer that behavior when it otherwise contravenes the DITA rules for topic groups (no navtitle, no referenced topic).

**Comment by robander** [comment]
@drmacro OK -- so given that, how would you see the TOC structure for this (assuming the navtitles here match what's in the topic)? And please note, I'm thinking of this as the edge case ... I'm happy with treating these the same as topicgroup when no parts have topics or titles...

```
<part collection-type="sequence">
  <chapter href="chap1.dita" navtitle="Good thing"/>
  <chapter href="chap2.dita" navtitle="Better thing"/>
</part>
<part href="ugly.dita" navtitle="This is icky">
  <chapter href="chap3.dita" navtitle="Best thing"/>
</part>
<part>
  <chapter href="chap4.dita" navtitle="Worst thing"/>
</part>
```

Currently, we generate this in the TOC, with pages at end of each:
```
Part 1.................
  Chapter 1. Good Thing.......
  Chapter 2. Better Thing.........
Part 2. This is icky........
  Chapter 3. Best Thing.......
Part 3..............
  Chapter 4. Worst thing.....
```

The bookmarks today are the same, except they drop the "Part X" and "Chapter X" for brevity, which is how we end up with the ugly empty bookmark for the first and last parts.

**Comment by drmacro** [comment]
There would be no title before chapter 1 and what is part 2 and 3 would be part 1 and 2

**Comment by robander** [comment]
So - we'd still have the empty "part" for what is today Part 3? 

TOC would become:
```
Chapter 1. Good Thing.......
Chapter 2. Better Thing.........
Part 1. This is icky........
  Chapter 3. Best Thing.......
Part 2..............
  Chapter 4. Worst thing.....
```

And bookmarks would become:
```
Good Thing.......
Better Thing.........
This is icky........
   Best Thing.......
Part 3
   Worst thing.....
```

@raducoravu and @jelovirt -- does that sound right, or do you have any other suggestions?

**Comment by raducoravu** [comment]
The part 3 in your example also does not have a navtitle so it should probably get treated exactly as the part 1 in the example (at least I do not see a reason for it to be treated differently).
So the TOC would be:

     Chapter 1. Good Thing.......
     Chapter 2. Better Thing.........
     Part 1. This is icky........
         Chapter 3. Best Thing.......
     Chapter 4. Worst thing.....

and the bookmarks:

    Good Thing.......
    Better Thing.........
    This is icky........
       Best Thing.......
    Worst thing.....


One thing about this issue is that it seems people did not complain about it much, it's been around for 3 years so it seems that maybe this usecase of using the "part" as an attributes container was something only specific to our user's guide. It depends of the development effort, if it would be a large one then the issue could just be closed and we'll see if other people will want the same thing in the future.

**Comment by robander** [comment]
> One thing about this issue is that it seems people did not complain about it much, it's been around for 3 years so it seems that maybe this usecase of using the "part" as an attributes container was something only specific to our user's guide.

Yes ... for me it seems like any of these are just "odd behavior for unusual source" so either of our TOC suggestions is probably fine, and better than what we have today. The only reason I was looking at this is because I thought I could take it with (relatively) small effort and it was in your list of priorities. 😉

**Comment by robander** [comment]
OK, I think I've got a fix for this one, because it's been bugging me, but it turned out a bit more complicated than I expected.

The `topicmerge` code explicitly creates a stub topic for every part. Initially it seemed like getting rid of that would be a simple approach. But, nope, it throws everything off because of code in the main process that just _expects_ such topics to exist for every part.

Instead I added a `@dita-ot:topicmerge` attribute that made it easier to recognize these empty `<part>` topics, and then skipped them when creating toc / bookmarks / page sequences. Seems to work well with some outlandish test cases I've created (part-container => real part => part-container => real part); also work with a map that only uses parts as containers.

Since it has little touches in several files, I'm not sure if this should go into a patch or wait for a full release - it may be possible the change in page sequences could throw off a plugin somewhere, if that plugin already handles this condition?
